### ISSUE TITLE
sql/multitenancy: incorrect results for interleaved join in tenant

### ISSUE THREAD

**Issue by asubiotto** [cockroachdb/cockroach#50050]
`testdata/logic_test/interleaved_join` fails when run with the `3node-tenant` config:
```
        testdata/logic_test/interleaved_join:174: SELECT * FROM parent1 JOIN child1 USING(pid1) WHERE pid1 >= 3 AND pid1 <= 5
        expected:
            pid1  pa1  cid1  ca1
            3     3    3     3
            3     3    43    43
            3     3    83    17
            3     3    123   57
            4     4    4     4
            4     4    44    44
            4     4    84    18
            4     4    124   58
            5     5    5     5
            5     5    45    45
            5     5    85    19
            5     5    125   59

        but found (query options: "rowsort,colnames" -> ignore the following ordering of rows) :
            pid1  pa1  cid1  ca1
```
Might be related to #50048

**Comment by asubiotto** [comment]
Tried to create a minimal repro with:
```
# LogicTest: 3node-tenant

statement ok
CREATE TABLE parent (pid1 INT PRIMARY KEY, pa1 INT, FAMILY (pid1), FAMILY (pa1))

statement ok
CREATE TABLE child (pid1 INT, cid1 INT, ca1 INT, PRIMARY KEY(pid1, cid1)) INTERLEAVE IN PARENT parent (pid1)

statement ok
INSERT INTO parent VALUES (1, 2)

statement ok
INSERT INTO child VALUES (1, 2, 3)

query IIII rowsort,colnames
SELECT * FROM parent JOIN child USING(pid1) WHERE pid1 >= 1 AND pid1 <= 5
----
pid1  pa1  cid1  ca1
1     2    2     3
```

But I'm now running into an `UnsupportedWithMultiTenancy` issue pointing back to this issue number caused by this usage of `NodeID`:
https://github.com/cockroachdb/cockroach/blob/fdf54ebc69c7d975f8f1dfbf28ced79ddeb83ee4/pkg/sql/distsql_plan_join.go#L272

@yuzefovich I'm not sure we want this, could you take a look at seeing if there's an alternative to reach the same goal (perhaps using the `DistSQLPlanner`'s `gatewayNodeID` field which is a cast sql instance id?) We can't disable interleaved joins in a multitenant environment.

**Comment by asubiotto** [comment]
I'm changing that line to `plan.GatewayNodeID = dsp.gatewayNodeID` for the meantime. If that's good enough, I'll merge that at the same time I merge a fix for this issue.

**Comment by asubiotto** [comment]
It looks like this issue has the same underlying cause as #50048 and will be fixed by #52562
